### PR TITLE
[Packets] Fix ALL Packets and ALL Packets correctly.

### PR DESCRIPTION
**Pull request by ghost**
This PR is created for fix all bugs about the Packets incorrect and add all encode and decode to them.

**My TODO list:**
1. [X] Part 1.
2. [X] Part 2.
3. [x] Part 3.
//4. [ ] Verify all and Remove unused imports.
//5. [ ] Reformat.
//6. [ ] https://github.com/sel-project/sel-utils/blob/master/xml/protocol/pocket134.xml -> Check and there AND add information about packets.

**Anyone can test changes and download JAR from: https://circleci.com/gh/NycuRO/Nukkit**
**ATENTION, PATCH-3**
**MASTER IS FOR PR GENERATOR**

### CODE REVIEW

**Comment by ghost** [comment]
@PikyCZ you can continue from there.

**Comment by PikyCZ** [comment]
why?

**Comment by ghost** [comment]
ok

File: src/main/java/cn/nukkit/network/protocol/ClientboundMapItemDataPacket.java [hunk 1]
```diff
@@ -10,15 +11,17 @@
...
     public int[] eids = new int[0];
 
     public long mapId;
     public int update;
     public byte scale;
     public int width;
     public int height;
-    public int offsetX;
-    public int offsetZ;
+    public int offsetX = 0;
```
Comment (ghost) [inline src/main/java/cn/nukkit/network/protocol/ClientboundMapItemDataPacket.java:23]: @Creeperface01 that fix lags.

**Comment by ghost** [comment]
@PikyCZ Hello, you continue or not?

**Comment by Snake1999** [comment]
:) Good job

**Comment by ghost** [comment]
@Snake1999 Probably i can continue and I. You want help me ? 👍

**Comment by ghost** [comment]
Translate from PM-MP branch 1.2 , it's not hard :D 

It's hard translate CommandBlockPacket and Inventory Packet, idk, but it's not same.

**Comment by larryTheCoder** [comment]
@NycuRO Its easy if you take a lot of care :D

**Comment by ghost** [comment]
@larryTheCoder :D

**Comment by ghost** [comment]
Note: PM-MP : Vecto3Obj
Nukkit: Vector3f

**Comment by ghost** [comment]
I'm fixed all errors, and updated to 1.2.0.22.

Anyone can test 👍

**Comment by ghost** [comment]
https://cdn.discordapp.com/attachments/346345325071695874/349983500922781696/Screenshot_2017-08-23-21-29-05.png

Works enter.

Bugs: 
1. Drops not work (@kvenatic97).
2. Bug duplication with PC Interface on Android.
3. Particles not work

**Comment by ghost** [comment]
AND CRASH BUG FIXED ,YEAH! @Snake1999 👍

**Comment by ghost** [comment]
image


I give this in console ,idk why.

**Comment by ghost** [comment]
pvp not work, get this error when hit 


image


when i'm die and respawn, get respawn in these position and get blocked in block.

**Comment by ghost** [comment]
maybe? https://github.com/pmmp/PocketMine-MP/commit/bb4808c23e9753622a3f0c989843950681b45ca6

File: src/main/java/cn/nukkit/network/protocol/ExplodePacket.java [hunk 2]
```diff
@@ -29,7 +30,24 @@ public DataPacket clean() {
...
+        this.y = v.y;
+        this.z = v.z;
+        this.radius = (float) (this.getVarInt() / 32);
 
+        int count = (int) this.getUnsignedVarInt();
+        for (int i = 0; i < count; ++i) {
+            x = y = z = 0.0f; // NULL
+            // $this->getSignedBlockPosition($x, $y, $z);
+            // Idk it's correct. Need check. @Anyone
+            BlockVector3 v = this.getBlockVector3();
```
Comment (ghost) [inline src/main/java/cn/nukkit/network/protocol/ExplodePacket.java]: remove this

**Comment by ghost** [comment]
@PikyCZ Gosh, yes!

**Comment by ghost** [comment]
@yescallop Good luck with translation from PM-MP, my work it's Packets 📦 

@PikyCZ give you Branch and me PR & File/Commits: https://github.com/pmmp/PocketMine-MP/pull/1311

**Comment by kvetinac97** [comment]
I'm going to test it

**Comment by ghost** [comment]
Before all, i verify and with https://github.com/sel-project/sel-utils/blob/master/xml/protocol/pocket134.xml

👍

**Comment by ghost** [comment]
@PikyCZ FINNALY. GOOD NIGHT ALL.

File: src/main/java/cn/nukkit/Player.java [hunk 3]
```diff
@@ -2242,11 +2242,11 @@ public void handleDataPacket(DataPacket packet) {
                         item = mobEquipmentPacket.item;
                         slot = Item.getCreativeItemIndex(item);
                     } else {
-                        item = this.inventory.getItem(mobEquipmentPacket.slot);
```
Comment (ghost) [inline src/main/java/cn/nukkit/Player.java:2245]: i think it's not all good. i check morning.

File: src/main/java/cn/nukkit/Player.java [hunk 4]
```diff
@@ -2242,11 +2242,11 @@ public void handleDataPacket(DataPacket packet) {
                         item = mobEquipmentPacket.item;
                         slot = Item.getCreativeItemIndex(item);
                     } else {
-                        item = this.inventory.getItem(mobEquipmentPacket.slot);
-                        slot = mobEquipmentPacket.slot;
+                        item = this.inventory.getItem(mobEquipmentPacket.inventorySlot);
+                        slot = mobEquipmentPacket.hotbarSlot;
```
Comment (ghost) [inline src/main/java/cn/nukkit/Player.java]: there sorry*

File: src/main/java/cn/nukkit/network/protocol/AvailableCommandsPacket.java [hunk 5]
```diff
@@ -18,18 +18,24 @@
...
+    /*public static final int ARG_TYPE_POSITION    = 0x0d; */
+    public static final int  ARG_TYPE_POSITION  = 0x13;
+    /*public static finaint nt ARG_TYPE_RAWTEXT = 0x10; */
+    public static final int  ARG_TYPE_RAWTEXT   = 0x16;
+    /*public static final int ARG_TYPE_TEXT    = 0x12; */
+    public static final int  ARG_TYPE_TEXT      = 0x18;
+    public static final int  ARG_TYPE_JSON      = 0x21;
+    public static final int  ARG_TYPE_COMMAND   = 0x28;
+    public static final int  ARG_FLAG_ENUM      = 0x2097152;
+    public static final int  ARG_FLAG_TEMPLATE  = 0x16777216;
```
Comment (ghost) [inline src/main/java/cn/nukkit/network/protocol/AvailableCommandsPacket.java]: All of there is from documentation. 👍

**Comment by ghost** [comment]
Wow, all type is now types, waw idk. ok ,now, gn.

File: src/main/java/cn/nukkit/network/protocol/AvailableCommandsPacket.java (hunk 5 above)
Comment (kvetinac97) [reply src/main/java/cn/nukkit/network/protocol/AvailableCommandsPacket.java]: 0x10 = 16 :D I guess  you didn't understand hexadecimal numbers

Comment (ghost) [reply src/main/java/cn/nukkit/network/protocol/AvailableCommandsPacket.java]: @kvetinac97 No, need search on google. 

I simple do "0x and number" =))
👍

**Comment by ghost** [comment]
* And add all recipes + creativeitems in .json, new feature on final.

**Comment by ghost** [comment]
@kvetinac97 Help?

**Comment by ghost** [comment]
@kvetinac97 i think it's my problem because all servers not work for me.

**Comment by ghost** [comment]
Anyone can test if work?

**Comment by kvetinac97** [comment]
I can enter, but it is spamming lots of errors into console about blockSkyLight

**Comment by ghost** [comment]
@kvetinac97 Can you show me?

**Comment by ghost** [comment]
@Snake1999 How does it look like before?
It's good? 👍

**Comment by ghost** [comment]
That's final @Snake1999 . You can accept @yescallop . Thanks for Nukkit but i'm leave from them.

**Comment by kvetinac97** [comment]
I will test

**Comment by kvetinac97** [comment]
Build fails

**Comment by ghost** [comment]
Where?

**Comment by ghost** [comment]
Now works @kvetinac97. Sorry, my issue.

**Comment by Snake1999** [comment]
Is this ready?

**Comment by ghost** [comment]
@Snake1999 Yes

**Comment by ghost** [comment]
But not updated to latest. I don't use nukkit in production, sorry, and that don't give me contribute :(

Sorry, but more owners not contribute to them to get good.

**Comment by Goodstuff4UonYT** [comment]
@NycuRO :c

**Comment by Goodstuff4UonYT** [comment]
@NycuRO Did you test this?
I cant start the server:

```
22:13:05 [ALERT] java.lang.NullPointerException
        at java.io.Reader.<init>(Reader.java:78)
        at java.io.InputStreamReader.<init>(InputStreamReader.java:113)
        at cn.nukkit.utils.Utils.readFile(Utils.java:63)
        at cn.nukkit.Server.<init>(Server.java:222)
        at cn.nukkit.Nukkit.main(Nukkit.java:68)


```

**Comment by Goodstuff4UonYT** [comment]
@NycuRO 
`
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: 2
        at java.lang.AbstractStringBuilder.charAt(AbstractStringBuilder.java:237)
        at java.lang.StringBuilder.charAt(StringBuilder.java:76)
        at jline.console.ConsoleReader.wcwidth(ConsoleReader.java:555)
        at jline.console.ConsoleReader.backspace(ConsoleReader.java:1057)
        at jline.console.ConsoleReader.backspaceAll(ConsoleReader.java:1043)
        at jline.console.ConsoleReader.resetPromptLine(ConsoleReader.java:3858)
        at cn.nukkit.command.CommandReader.removePromptLine(CommandReader.java:120)
        at cn.nukkit.Nukkit.main(Nukkit.java:93)
`

**Comment by Goodstuff4UonYT** [comment]
@NycuRO @Snake1999 @PikyCZ Cant brake blocks (In gamemode 0 or 1). There is no block brake animation, like if I am in gamemode 2.
There are also shown bubbles the whole time.
And I am flying, but cant controll the high (in gamemode 0!)
Nothing is working :c

**Comment by Goodstuff4UonYT** [comment]
whatsapp image 2017-09-16 at 23 46 32

**Comment by ghost** [comment]
@Goodstuff4UonYT It's not by packets bug.

**Comment by JackboyPlay** [comment]
What should be changed that this could be merged?

**Comment by boybook** [comment]
That bug may caused by AdventureSettingsPacket.
And maybe there are something wrong in entity metadata encoding.

**Comment by boybook** [comment]
We must speed up the upgrade process!

**Comment by ghost** [comment]
@boybook I'm stopped working. Thanks for all support on nukkit.

Good luck with more.

**Comment by Goodstuff4UonYT** [comment]
@boybook https://github.com/Creeperface01/Nukkit/tree/GT1.2

**Comment by Goodstuff4UonYT** [comment]
@boybook 
https://github.com/Creeperface01/Nukkit/tree/GT1.2 version of Nukkit 1.2 :

- [ ] - if you teleport somebody he falls through the blocks often (Was not in 1.1!)
- [ ] - if you hit the ground with a item sometimes the client crashs
- [x] - Floating Text Particle does not work
- [ ] - Win 10 disconnects often, If you pick up an item, or walk into water etc.
- [ ] - If you hit entities which were placed in 1.1, some entities: no entity damage event is called.
- [ ] - Does not work anymore:

```
player.getAdventureSettings().set(Type.ALLOW_FLIGHT, true);
player.getAdventureSettings().update();
```

- [x] - https://github.com/Creeperface01/Nukkit/tree/GT1.2 must merge into master (conflicts!)
- [ ] - Sounds are broken. (Hit a noteblock, you can't hear anything)

**Pull request closed without merging**